# Zadanie VIII

In [ ]:
conda install pandas

In [ ]:
conda install scikit-learn

In [ ]:
conda install matplotlib

In [35]:
import pandas as pd

In [36]:
from sklearn.feature_extraction.text import TfidfTransformer

In [37]:
import numpy as np

In [38]:
from sklearn.feature_extraction.text import CountVectorizer

In [39]:
import matplotlib.pyplot as plt

In [56]:
from sklearn.linear_model import LogisticRegression

### Zadanie

In [106]:
data = pd.read_csv('https://www.fizyka.umk.pl/~grochu/wdm/files/spam_train.csv')
datav1 = data.replace('ham', 0)
datav1 = datav1.replace('spam', 1)

In [107]:
data2 = pd.read_csv('https://www.fizyka.umk.pl/~grochu/wdm/files/spam_test.csv')
datav2 = data2.replace('ham', 0)
datav2 = datav2.replace('spam', 1)

In [132]:
import re
def preprocessor(text):
    text = re.sub('<[^>]*>', '', text)
    text = re.sub('[\W]+', ' ', text.lower())
    return text

In [109]:
X_datav1 = datav1.iloc[:,1]

In [110]:
Y_datav1 = datav1.iloc[:,0]

In [112]:
X_datav1 = X_datav1.apply(preprocessor)

In [113]:
X_datav1

0       go until jurong point crazy available only in ...
1                                ok lar joking wif u oni 
2       free entry in 2 a wkly comp to win fa cup fina...
3            u dun say so early hor u c already then say 
4       nah i don t think he goes to usf he lives arou...
                              ...                        
3995    we ll you pay over like lt gt yrs so its not t...
3996    bored housewives chat n date now 0871750 77 11...
3997    we tried to call you re your reply to our sms ...
3998                        k k when will you give treat 
3999    this is the 2nd time we have tried to contact ...
Name: SMS, Length: 4000, dtype: object

In [114]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [115]:
tfidf = TfidfVectorizer(stop_words='english')

X_tfidf_train = tfidf.fit_transform(X_datav1)

In [116]:
print('Wielkość słownika: ', len(tfidf.get_feature_names()))

Wielkość słownika:  7036


In [117]:
lr = LogisticRegression(C=10, solver='lbfgs', random_state=0)
lr.fit(X_tfidf_train, Y_datav1)

LogisticRegression(C=10, random_state=0)

In [121]:
X_datav2 = datav2.iloc[:,1]

In [122]:
Y_datav2 = datav2.iloc[:,0]

In [126]:
X_datav2 = X_datav2.apply(preprocessor)

In [127]:
X_datav2

4000    he s just gonna worry for nothing and he won t...
4001    did you get any gift this year i didnt get any...
4002    somewhere out there beneath the pale moon ligh...
4003    well there s a pattern emerging of my friends ...
4004     ow u dey i paid 60 400thousad i told u would ...
                              ...                        
5567    this is the 2nd time we have tried 2 contact u...
5568                will ì_ b going to esplanade fr home 
5569    pity was in mood for that so any other suggest...
5570    the guy did some bitching but i acted like i d...
5571                            rofl its true to its name
Name: SMS, Length: 1572, dtype: object

In [128]:
X_tfidf_test = tfidf.transform(X_datav2)

In [129]:
y_pred = lr.predict(X_tfidf_test)

In [118]:
from sklearn.metrics import accuracy_score

In [130]:
print('Accuracy score: ', accuracy_score(Y_datav2, y_pred))

Accuracy score:  0.9790076335877863


In [131]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(Y_datav2, y_pred)
cc = pd.DataFrame(cm, columns=['negative', 'positive'], index=['negative', 'positive'])
cc.columns.name = 'prediction'
cc.index.name = 'real'
cc

prediction  negative  positive
real                          
negative        1356         4
positive          29       183

### Okazuje się, że skuteczność jest na bardzo wysokim poziomie - aż 97.9%